### ISSUE TITLE
Some isinstance-guarded return statements do not typecheck against bounded type variables

### ISSUE THREAD

**Issue by KevOrr** [microsoft/pyright#2807]
**Describe the bug**
Following up from the conversation in https://github.com/microsoft/pyright/issues/2806#issuecomment-1005212077, it seems that some functions mess up the `isinstance`-guarded return type narrowing.

**To Reproduce**
Typecheck the following snippet:
```py
def foo(s: AnyStr, y: Optional[AnyStr] = None) -> AnyStr:
    if isinstance(s, str):
        if y is None:
            pass
        return 'asdf'
    else:
        raise NotImplementedError
```
```
  error: Expression of type "Literal['asdf']" cannot be assigned to return type "AnyStr@foo"
    Type "Literal['asdf']" cannot be assigned to type "AnyStr@foo" (reportGeneralTypeIssues)
```

Thinking this was just a bug with `Literal`, I tried the following:
```py
def foo(s: AnyStr, y: Optional[AnyStr] = None) -> AnyStr:
    if isinstance(s, str):
        if y is None:
            pass
        x: str = str('asdf')
        return x
    else:
        raise NotImplementedError
```
```
  error: Expression of type "str" cannot be assigned to return type "AnyStr@foo"
    Type "str" cannot be assigned to type "AnyStr@foo" (reportGeneralTypeIssues)
```

However, if the `if y ... pass` block is taken out, both versions of `foo` typecheck.

**VS Code extension or command-line**
CLI and as an LSP server. 1.1.203

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.204, which I just published. It will also be included in the next release of pylance.
